### PR TITLE
[Datadog] Onboard Datadog to CLI

### PR DESCRIPTION
**Pull request by limingu**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Datadog

File: src/datadog/azext_datadog/generated/commands.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        g.custom_show_command('show', 'datadog_tag_rule_show')
+        g.custom_command('create', 'datadog_tag_rule_create')
+        g.custom_command('update', 'datadog_tag_rule_update')
+
+    from azext_datadog.generated._client_factory import cf_single_sign_on_configuration
+    datadog_single_sign_on_configuration = CliCommandType(
+        operations_tmpl='azext_datadog.vendored_sdks.datadog.operations._single_sign_on_configuration_operations#Single'
+        'SignOnConfigurationOperations.{}',
+        client_factory=cf_single_sign_on_configuration)
+    with self.command_group('datadog single-sign-on-configuration', datadog_single_sign_on_configuration,
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/commands.py:89]: Consider using `datadog sso-config` to make the command short.

File: src/datadog/azext_datadog/generated/_help.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+"""
+
+helps['datadog refresh-set-password'] = """
+    type: group
+    short-summary: datadog refresh-set-password
+"""
+
+helps['datadog refresh-set-password get'] = """
+    type: command
+    short-summary: "Refresh the set password link and return a latest one."
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_help.py:179]: The command name `datadog refresh-set-password get` is a bit difficult to understand and CLI command name needs to follow `noun [noun] ... verb` format.

File: src/datadog/azext_datadog/generated/_help.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    examples:
+      - name: TagRules_Get
+        text: |-
+               az datadog tag-rule show --monitor-name "myMonitor" --resource-group "myResourceGroup" --rule-set-name \
+"default"
+"""
+
+helps['datadog tag-rule create'] = """
+    type: command
+    short-summary: "Create or update a tag rule set for a given monitor resource."
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_help.py:212]: The short summary needs to remove `or update` as this is for the `create` command.

File: src/datadog/azext_datadog/generated/_help.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        text: |-
+               az datadog tag-rule create --monitor-name "myMonitor" --log-rules-filtering-tags name="Environment" \
+action="Include" value="Prod" --log-rules-filtering-tags name="Environment" action="Exclude" value="Dev" \
+--log-rules-send-aad-logs false --log-rules-send-resource-logs true --log-rules-send-subscription-logs true \
+--resource-group "myResourceGroup" --rule-set-name "default"
+"""
+
+helps['datadog tag-rule update'] = """
+    type: command
+    short-summary: "Create or update a tag rule set for a given monitor resource."
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_help.py:250]: The short summary needs to remove `Create or` as this is for the `update` command.

File: src/datadog/azext_datadog/generated/_params.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        c.argument('resource_group_name', resource_group_name_type)
+        c.argument('monitor_name', options_list=['--name', '-n', '--monitor-name'], type=str, help='Monitor resource '
+                   'name')
+        c.argument('tags', tags_type)
+        c.argument('location', arg_type=get_location_type(self.cli_ctx),
+                   validator=get_default_location_from_resource_group)
+        c.argument('identity_type', arg_type=get_enum_type(['SystemAssigned', 'UserAssigned']), help='Identity type')
+        c.argument('provisioning_state', arg_type=get_enum_type(['Accepted', 'Creating', 'Updating', 'Deleting', ''
+                                                                 'Succeeded', 'Failed', 'Canceled', 'Deleted', ''
+                                                                 'NotSpecified']), help='Provisioning state')
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_params.py:78]: The extra `''` in the end is not necessary. @qiaozha to check formatting.

File: src/datadog/azext_datadog/generated/_params.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        c.argument('monitoring_status', arg_type=get_enum_type(['Enabled', 'Disabled']), help='Flag specifying if the '
+                   'resource monitoring is enabled or disabled.')
+        c.argument('marketplace_subscription_status', arg_type=get_enum_type(['Active', 'Suspended']), help='Flag '
+                   'specifying the Marketplace Subscription Status of the resource. If payment is not made in time, '
+                   'the resource will go in Suspended state.')
+        c.argument('datadog_organization_properties', action=AddDatadogOrganizationProperties, nargs='*', help='')
+        c.argument('user_info', action=AddUserInfo, nargs='*', help='')
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_params.py:85]: Please add help message for these 2 properties.

File: src/datadog/azext_datadog/generated/_help.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+            email-address: Email of the user used by Datadog for contacting them if needed
+            phone-number: Phone number of the user used by Datadog for contacting them if needed
+    examples:
+      - name: Monitors_Create
+        text: |-
+               az datadog monitor create --name "myMonitor" --sku-name "myMonitor" --location "West US 2" \
+--datadog-organization-properties name="myOrg" enterprise-app-id="00000000-0000-0000-0000-000000000000" \
+linking-auth-code="someAuthCode" linking-client-id="00000000-0000-0000-0000-000000000000" subscription="pro" \
+--monitoring-status "Enabled" --user-info name="Alice" email-address="alice@microsoft.com" phone-number="123-456-7890" \
+--sku-name "drawdown_testing_20200904_Monthly" --tags Environment="Dev" --resource-group "myResourceGroup"
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_help.py:138]: `--sku-name` appear twice in this example with different values.

File: src/datadog/azext_datadog/generated/_params.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        c.argument('sku_name', type=str, help='Name of the SKU.')
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_params.py:86]: The `sku_name` can be any string or there is a list of choices?

File: src/datadog/azext_datadog/generated/_help.py (hunk 2 above)
Comment (limingu) [reply src/datadog/azext_datadog/generated/_help.py:179]: Updated.

File: src/datadog/azext_datadog/generated/_help.py (hunk 3 above)
Comment (limingu) [reply src/datadog/azext_datadog/generated/_help.py:212]: Updated.

File: src/datadog/azext_datadog/generated/_help.py (hunk 4 above)
Comment (limingu) [reply src/datadog/azext_datadog/generated/_help.py:250]: Updated.

File: src/datadog/azext_datadog/generated/_params.py (hunk 6 above)
Comment (limingu) [reply src/datadog/azext_datadog/generated/_params.py:85]: Updated.

File: src/datadog/azext_datadog/generated/_help.py (hunk 7 above)
Comment (limingu) [reply src/datadog/azext_datadog/generated/_help.py:138]: Updated.

File: src/datadog/azext_datadog/generated/_params.py (hunk 8 above)
Comment (limingu) [reply src/datadog/azext_datadog/generated/_params.py:86]: The sku_name has some choices. But it is not a fixed list as Datadog can add some new sku without letting us know. So we keep it as a string without any validation or limitation.

File: src/datadog/azext_datadog/generated/_params.py (hunk 5 above)
Comment (limingu) [reply src/datadog/azext_datadog/generated/_params.py:78]: I think this is auto generated. '' is used to end this line. I see this formatting in other places as well.

File: src/datadog/README.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Microsoft Azure CLI 'datadog' Extension
+==========================================
+
+This package is for the 'datadog' extension.
+i.e. 'az datadog'
```
[the remark this answers is not in the document]
Comment (limingu) [reply src/datadog/README.md:5]: As Datadog service hasn't been in public preview, there is no public document or link for Datadog service. Can we add it later once we have the public document link?

File: src/datadog/azext_datadog/generated/_params.py (hunk 5 above)
Comment (qiaozha) [reply src/datadog/azext_datadog/generated/_params.py]: @changlong-liu Could you take a look? Thanks

Comment (changlong-liu) [reply src/datadog/azext_datadog/generated/_params.py]: Yes, it's a codegen issue.

Hi @limingu , to fix this PR, would you change code manually for short-term solution? 
I just add an issue in codegen for tracking and fixing it: https://github.com/Azure/autorest.az/issues/591

File: src/datadog/azext_datadog/generated/_params.py (hunk 8 above)
Comment (haroldrandom) [reply src/datadog/azext_datadog/generated/_params.py:87]: it would be better to let the user know which SDK number they are able to fill

**Review by haroldrandom — APPROVED** [review]

**Review by haroldrandom — APPROVED** [review]
Able to merge once other reviewers approve

File: src/datadog/azext_datadog/generated/_help.py [hunk 10]
```diff
@@ -169,18 +169,18 @@
                az datadog monitor wait --name "myMonitor" --resource-group "myResourceGroup" --deleted
 """
 
-helps['datadog refresh-set-password'] = """
+helps['datadog set-passward-link'] = """
```
Comment (fengzhou-msft) [inline src/datadog/azext_datadog/generated/_help.py:172]: typo: `passward`?

Comment (limingu) [reply src/datadog/azext_datadog/generated/_help.py]: Fixed. Thanks!

**Review by fengzhou-msft — APPROVED** [review]

**Pull request merged**